File: nRF Toolbox/Utilities/CharacteristicReader.m
/*
 * Copyright (c) 2015, Nordic Semiconductor
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without modification, are permitted provided that the following conditions are met:
 *
 * 1. Redistributions of source code must retain the above copyright notice, this list of conditions and the following disclaimer.
 *
 * 2. Redistributions in binary form must reproduce the above copyright notice, this list of conditions and the following disclaimer in the
 * documentation and/or other materials provided with the distribution.
 *
 * 3. Neither the name of the copyright holder nor the names of its contributors may be used to endorse or promote products derived from this
 * software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
 * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
 * HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
 * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON
 * ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE
 * USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

#import "CharacteristicReader.h"


typedef enum {
    MDER_S_POSITIVE_INFINITY = 0x07FE,
    MDER_S_NaN = 0x07FF,
    MDER_S_NRes = 0x0800,
    MDER_S_RESERVED_VALUE = 0x0801,
    MDER_S_NEGATIVE_INFINITY = 0x0802
} ReservedSFloatValues;
static const UInt32 FIRST_S_RESERVED_VALUE = MDER_S_POSITIVE_INFINITY;

typedef enum {
    MDER_POSITIVE_INFINITY = 0x007FFFFE,
    MDER_NaN = 0x007FFFFF,
    MDER_NRes = 0x00800000,
    MDER_RESERVED_VALUE = 0x00800001,
    MDER_NEGATIVE_INFINITY = 0x00800002
} ReservedFloatValues;
static const UInt32 FIRST_RESERVED_VALUE = MDER_POSITIVE_INFINITY;

static const double reserved_float_values[5] = {INFINITY, NAN, NAN, NAN, -INFINITY};

@implementation CharacteristicReader


+ (UInt8)readUInt8Value:(uint8_t **)p_encoded_data
{
    return *(*p_encoded_data)++;
}

+ (SInt8)readSInt8Value:(uint8_t **)p_encoded_data
{
    return *(*p_encoded_data)++;
}

+ (UInt16)readUInt16Value:(uint8_t **)p_encoded_data
{
    UInt16 value = (UInt16) CFSwapInt16LittleToHost(*(uint16_t*)*p_encoded_data);
    *p_encoded_data += 2;
    return value;
}

+ (SInt16)readSInt16Value:(uint8_t **)p_encoded_data
{
    SInt16 value = (SInt16) CFSwapInt16LittleToHost(*(uint16_t*)*p_encoded_data);
    *p_encoded_data += 2;
    return value;
}

+ (UInt32)readUInt32Value:(uint8_t **)p_encoded_data
{
    UInt32 value = (UInt32) CFSwapInt32LittleToHost(*(uint32_t*)*p_encoded_data);
    *p_encoded_data += 4;
    return value;
}

+ (SInt32)readSInt32Value:(uint8_t **)p_encoded_data
{
    SInt32 value = (SInt32) CFSwapInt32LittleToHost(*(uint32_t*)*p_encoded_data);
    *p_encoded_data += 4;
    return value;
}

+ (Float32)readSFloatValue:(uint8_t **)p_encoded_data
{
    UInt16 tempData = CFSwapInt16LittleToHost(*(uint16_t*)*p_encoded_data);
    
    SInt16 mantissa = tempData & 0x0FFF;
    SInt8 exponent = tempData >> 12;
    
    if (exponent >= 0x0008) {
        exponent = -((0x000F + 1) - exponent);
    }
    
    Float32 output = 0;
    
    if (mantissa >= FIRST_S_RESERVED_VALUE && mantissa <= MDER_S_NEGATIVE_INFINITY)
    {
        output = reserved_float_values[mantissa - FIRST_S_RESERVED_VALUE];
    }
    else
    {
        if (mantissa >= 0x0800)
        {
            mantissa = -((0x0FFF + 1) - mantissa);
        }
        double magnitude = pow(10.0f, exponent);
        output = (mantissa * magnitude);
    }
    
    *p_encoded_data += 2;
    return output;
}

+(Float32)readFloatValue:(uint8_t **)p_encoded_data
{
    SInt32 tempData = (SInt32) CFSwapInt32LittleToHost(*(uint32_t*)*p_encoded_data);
    
    SInt32 mantissa = tempData & 0xFFFFFF;
    SInt8 exponent = tempData >> 24;
    Float32 output = 0;
    
    if (mantissa >= FIRST_RESERVED_VALUE && mantissa <= MDER_NEGATIVE_INFINITY)
    {
        output = reserved_float_values[mantissa - FIRST_RESERVED_VALUE];
    }
    else
    {
        if (mantissa >= 0x800000)
        {
            mantissa = -((0xFFFFFF + 1) - mantissa);
        }
        double magnitude = pow(10.0f, exponent);
        output = (mantissa * magnitude);
    }
    
    *p_encoded_data += 4;
    return output;
}

+(NSDate *)readDateTime:(uint8_t **)p_encoded_data
{
    uint16_t year = [CharacteristicReader readUInt16Value:p_encoded_data];
    uint8_t month = [CharacteristicReader readUInt8Value:p_encoded_data];
    uint8_t day = [CharacteristicReader readUInt8Value:p_encoded_data];
    uint8_t hour = [CharacteristicReader readUInt8Value:p_encoded_data];
    uint8_t min = [CharacteristicReader readUInt8Value:p_encoded_data];
    uint8_t sec = [CharacteristicReader readUInt8Value:p_encoded_data];
    
    NSString * dateString = [NSString stringWithFormat:@"%d %d %d %d %d %d", year, month, day, hour, min, sec];
    
    NSDateFormatter *dateFormat = [[NSDateFormatter alloc] init];
    [dateFormat setDateFormat: @"yyyy MM dd HH mm ss"];
    return  [dateFormat dateFromString:dateString];
}

+(Nibble)readNibble:(uint8_t **)p_encoded_data
{
    Nibble nibble;
    nibble.value = [CharacteristicReader readUInt8Value:p_encoded_data];
    
    return nibble;
}

@end

File: nRF Toolbox/Utilities/CharacteristicReader.swift
//
//  CharacteristicReader.swift
//  nRF Toolbox
//
//  Created by Mostafa Berg on 18/05/16.
//  Copyright © 2016 Nordic Semiconductor. All rights reserved.
//

import Foundation

struct Nibble {
    var first  : UInt8
    var second : UInt8
}

enum ReservedSFloatValues : Int16 {
    case positiveInfinity = 0x07FE
    case nan = 0x07FF
    case nres = 0x0800
    case reserved = 0x0801
    case negativeInfinity = 0x0802
    
    static let firstReservedValue = ReservedSFloatValues.positiveInfinity
}

enum ReservedFloatValues : UInt32 {
    case positiveInfinity = 0x007FFFFE
    case nan = 0x007FFFFF
    case nres = 0x00800000
    case reserved = 0x00800001
    case negativeInfinity = 0x00800002
    
    static let firstReservedValue = ReservedFloatValues.positiveInfinity
}

extension Double {
    static var reservedValues: [Double] {
        [.infinity, .nan, .nan, .nan, -.infinity]
    }
}

struct CharacteristicReader {

    static func readUInt8Value(ptr aPointer : inout UnsafeMutablePointer<UInt8>) -> UInt8 {
        let val = aPointer.pointee
        aPointer = aPointer.successor()
        return val
    }
    
    static func readSInt8Value(ptr aPointer : UnsafeMutablePointer<UInt8>) -> Int8 {
        Int8(aPointer.successor().pointee)
    }

    static func readUInt16Value(ptr aPointer : inout UnsafeMutablePointer<UInt8>) -> UInt16 {
        let anUInt16Pointer = UnsafeMutablePointer<UInt16>(OpaquePointer(aPointer))
        let val = CFSwapInt16LittleToHost(anUInt16Pointer.pointee)
        aPointer = aPointer.advanced(by: 2)
        return val
    }
    
    static func readSInt16Value(ptr aPointer : inout UnsafeMutablePointer<UInt8>) -> Int16 {
        let anInt16Pointer = UnsafeMutablePointer<Int16>(OpaquePointer(aPointer))
        let val = CFSwapInt16LittleToHost(UnsafeMutablePointer<UInt16>(OpaquePointer(anInt16Pointer)).pointee)
        aPointer = aPointer.advanced(by: 2)
        return Int16(val)
    }
    
    static func readUInt32Value(ptr aPointer : inout UnsafeMutablePointer<UInt8>) -> UInt32 {
        let anInt32Pointer = UnsafeMutablePointer<UInt32>(OpaquePointer(aPointer))
        let val = CFSwapInt32LittleToHost(UnsafeMutablePointer<UInt32>(OpaquePointer(anInt32Pointer)).pointee)
        aPointer = aPointer.advanced(by: 4)
        return UInt32(val)
    }
    
    static func readSInt32Value(ptr aPointer : inout UnsafeMutablePointer<UInt8>) -> Int32 {
        let anInt32Pointer = UnsafeMutablePointer<UInt32>(OpaquePointer(aPointer))
        let val  = CFSwapInt32LittleToHost(UnsafeMutablePointer<UInt32>(OpaquePointer(anInt32Pointer)).pointee)
        aPointer = aPointer.advanced(by: 4)
        return Int32(val)
    }
    
    static func readSFloat(_ data: Data, offset: Int) -> Float32 {
        let tempData: UInt16 = data.read(fromOffset: offset)
        var mantissa = Int16(tempData & 0x0FFF)
        var exponent = Int8(tempData >> 12)
        if exponent >= 0x0008 {
            exponent = -( (0x000F + 1) - exponent )
        }
        
        var output : Float32 = 0
        
        if mantissa >= ReservedSFloatValues.firstReservedValue.rawValue && mantissa <= ReservedSFloatValues.negativeInfinity.rawValue {
            output = Float32(Double.reservedValues[Int(mantissa - ReservedSFloatValues.firstReservedValue.rawValue)])
        } else {
            if mantissa > 0x0800 {
                mantissa = -((0x0FFF + 1) - mantissa)
            }
            let magnitude = pow(10.0, Double(exponent))
            output = Float32(mantissa) * Float32(magnitude)
        }
        
        return output
    }
    
    static func readSFloatValue(ptr aPointer : inout UnsafeMutablePointer<UInt8>) -> Float32 {
        let tempData = CFSwapInt16LittleToHost(UnsafeMutablePointer<UInt16>(OpaquePointer(aPointer)).pointee)
        var mantissa = Int16(tempData & 0x0FFF)
        var exponent = Int8(tempData >> 12)
        if exponent >= 0x0008 {
            exponent = -((0x000F + 1) - exponent)
        }

        var output : Float32 = 0
        
        if mantissa >= ReservedSFloatValues.firstReservedValue.rawValue && mantissa <= ReservedSFloatValues.negativeInfinity.rawValue {
            output = Float32(Double.reservedValues[Int(mantissa - ReservedSFloatValues.firstReservedValue.rawValue)])
        } else {
            if mantissa > 0x0800 {
                mantissa = -((0x0FFF + 1) - mantissa)
            }
            let magnitude = pow(10.0, Double(exponent))
            output = Float32(mantissa) * Float32(magnitude)
        }
        aPointer = aPointer.advanced(by: 2)
        
        return output
    }
    
    static func readFloatValue(ptr aPointer : inout UnsafeMutablePointer<UInt8>) -> Float32 {
        let tempData = CFSwapInt32LittleToHost(UnsafeMutablePointer<UInt32>(OpaquePointer(aPointer)).pointee)
        var mantissa = Int32(tempData & 0x00FFFFFF)
        let exponent = Int8(bitPattern: UInt8(tempData >> 24))
        
        var output : Float32 = 0
        
        if mantissa >= Int32(ReservedFloatValues.firstReservedValue.rawValue) && mantissa <= Int32(ReservedFloatValues.negativeInfinity.rawValue) {
            output = Float32(Double.reservedValues[Int(mantissa - Int32(ReservedSFloatValues.firstReservedValue.rawValue))])
        } else {
            if mantissa >= 0x800000 {
                mantissa = -((0xFFFFFF + 1) - mantissa)
            }
            let magnitude = pow(10.0, Double(exponent))
            output = Float32(mantissa) * Float32(magnitude)
        }
        
        aPointer = aPointer.advanced(by: 4)
        return output
    }
    
    static func readDateTime(ptr aPointer : inout UnsafeMutablePointer<UInt8>) -> Date {
        let year  = CharacteristicReader.readUInt16Value(ptr: &aPointer)
        let month = CharacteristicReader.readUInt8Value(ptr: &aPointer)
        let day   = CharacteristicReader.readUInt8Value(ptr: &aPointer)
        var hour  = CharacteristicReader.readUInt8Value(ptr: &aPointer)
        let min   = CharacteristicReader.readUInt8Value(ptr: &aPointer)
        let sec   = CharacteristicReader.readUInt8Value(ptr: &aPointer)
        
        let dateFormatter = DateFormatter()

        var dateString : String
        if using12hClockFormat()  == true {
            var merediumString :String = "am"
            if (hour > 12) {
                hour = hour - 12;
                merediumString = "pm"
            }
            dateString = String(format: "%d %d %d %d %d %d %@", year, month, day, hour, min, sec, merediumString)
            dateFormatter.dateFormat = "yyyy MM dd HH mm ss a"
        }else{
            dateString = String(format: "%d %d %d %d %d %d", year, month, day, hour, min, sec)
            dateFormatter.dateFormat = "yyyy MM dd HH mm ss"
        }
        return dateFormatter.date(from: dateString)!
    }
    
    static func readNibble(ptr aPointer : inout UnsafeMutablePointer<UInt8>) -> Nibble {
        let value = CharacteristicReader.readUInt8Value(ptr: &aPointer)
        let nibble = Nibble(first: value & 0xF, second: value >> 4)
        return nibble
    }
    
    static func using12hClockFormat() -> Bool {
        
        let formatter = DateFormatter()
        formatter.locale = Locale.current
        formatter.dateStyle = .none
        formatter.timeStyle = .short
        
        let dateString = formatter.string(from: Date())
        let amRange = dateString.range(of: formatter.amSymbol)
        let pmRange = dateString.range(of: formatter.pmSymbol)
        
        return !(pmRange == nil && amRange == nil)
    }
}
